core/src/main/java/org/infinispan/util/concurrent/NonBlockingOrderer.java

```diff
@@ -0,0 +1,67 @@
+package org.infinispan.util.concurrent;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * Ordering construct allowing concurrent operations that wish to do operations upon the same key to wait until
+ * the most recently registered operation is complete in a non blocking way.
+ */
+public class NonBlockingOrderer {
+   private final static Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private final static boolean trace = log.isTraceEnabled();
+
+   private final ConcurrentMap<Object, CompletionStage<OPERATION>> objectStages = new ConcurrentHashMap<>();
+
+   public enum OPERATION {
+      READ,
+      REMOVE,
+      WRITE
+   }
+
+   /**
+    * Registers the provided Stage to be next in line to do an operation on behalf of the provided key.
+    * Returns a different Stage that when complete signals that this operation should continue or null if there
+    * is no wait required.
+    * @param key delineating identifier for an operation
+    * @param register stage to register for others to wait upon for future registrations
+    * @return stage that signals when the operation that is registering its own future may continue or null if nothing
+    *         to wait on
+    */
+   public CompletionStage<OPERATION> orderOn(Object key, CompletionStage<OPERATION> register) {
```

Comment:
The problem with using *Async methods is if the Stage is already complete it will still use another thread to continue execution, even though we could very well have used the current one. Although I am kinda thinking I may want to go back the other way and use *Async as more often than not they won't be completed at that exact point.

> It also seems "old-scool" to pass a CompletionStage in, I would have expected to pass a Function<OPERATION, CompletionStage<OPERATION>> instead.

Not sure I am following what you mean here.